### ISSUE TITLE
Error when compiling

### ISSUE THREAD

**Issue by michaellatham77** [haddocking/disvis#23]
Hi,

I'm getting the following errors when compiling

`src/_extensions.c:193:5: error: implicit declaration of function 'PyArray_XDECREF_ERR' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
PyArray_XDECREF_ERR(py_out);`

`src/_extensions.c:208:12: error: implicit declaration of function 'Py_InitModule' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
(void) Py_InitModule("_extensions", mymethods);`

I am compiling on a M1 Mac running OS 12.6 (Monterey) using python 3.9. 

Any assistance resolving this issue would be appreciated.

Cheers,
Mike

**Comment by rvhonorato** [comment]
Thanks for reporting it @michaellatham77, please share what is the output of the following command:

```bash
python3.9 -m venv .venv
source .venv/bin/activate
pip install numpy pyfftw
pip install git+https://github.com/haddocking/disvis.git@v2.1
```

**Comment by michaellatham77** [comment]
Dear Rodrigo,

Here is the output from the listed commands...

```
latha070% python3 -m venv .venv
latha070%
```

I don't have a `python3.9` command. `python3` does the same thing on my system.

```
latha070% source .venv/bin/activate
Badly placed ()'s.
```
```
latha070% pip install numpy pyfftw
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: numpy in /Users/latha070/Library/Python/3.9/lib/python/site-packages (1.23.4)
Collecting pyfftw
  Downloading pyFFTW-0.13.1.tar.gz (114 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.4/114.4 kB 3.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: pyfftw
  Building wheel for pyfftw (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for pyfftw (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [41 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-10.9-universal2-3.9
      creating build/lib.macosx-10.9-universal2-3.9/pyfftw
      copying pyfftw/config.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw
      copying pyfftw/_version.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw
      copying pyfftw/__init__.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw
      creating build/lib.macosx-10.9-universal2-3.9/pyfftw/builders
      copying pyfftw/builders/builders.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/builders
      copying pyfftw/builders/__init__.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/builders
      copying pyfftw/builders/_utils.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/builders
      creating build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/cache.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/__init__.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/scipy_fft.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/dask_fft.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/numpy_fft.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/scipy_fftpack.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      copying pyfftw/interfaces/_utils.py -> build/lib.macosx-10.9-universal2-3.9/pyfftw/interfaces
      UPDATING build/lib.macosx-10.9-universal2-3.9/pyfftw/_version.py
      set build/lib.macosx-10.9-universal2-3.9/pyfftw/_version.py to '0.13.1'
      running build_ext
      DEBUG:__main__:Link FFTW dynamically
      DEBUG:__main__:Compiler include_dirs: ['/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9']
      DEBUG:__main__:3.9.6 (default, Oct 18 2022, 12:41:40)
      [Clang 14.0.0 (clang-1400.0.29.202)]
      DEBUG:__main__:Sniffer include_dirs: ['/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-install-69kpg58x/pyfftw_de1d4c43bd094477b4060acd9ebf60c5/include', '/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-install-69kpg58x/pyfftw_de1d4c43bd094477b4060acd9ebf60c5/pyfftw', '/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-build-env-9j0btfec/overlay/lib/python3.9/site-packages/numpy/core/include', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include']
      DEBUG:__main__:objects: []
      DEBUG:__main__:libraries: []
      DEBUG:__main__:include dirs: ['/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-install-69kpg58x/pyfftw_de1d4c43bd094477b4060acd9ebf60c5/include', '/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-install-69kpg58x/pyfftw_de1d4c43bd094477b4060acd9ebf60c5/pyfftw', '/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-build-env-9j0btfec/overlay/lib/python3.9/site-packages/numpy/core/include', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include']
      /var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pyfftw-56f85mvn/None.c:1:10: fatal error: 'fftw3.h' file not found
      #include <fftw3.h>
               ^~~~~~~~~
      1 error generated.
      WARNING:__main__:Compilation error: command '/usr/bin/clang' failed with exit code 1
      DEBUG:__main__:clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration -I/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-install-69kpg58x/pyfftw_de1d4c43bd094477b4060acd9ebf60c5/include -I/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-install-69kpg58x/pyfftw_de1d4c43bd094477b4060acd9ebf60c5/pyfftw -I/private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-build-env-9j0btfec/overlay/lib/python3.9/site-packages/numpy/core/include -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -c /var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pyfftw-56f85mvn/None.c -o /var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pyfftw-56f85mvn/None.o
      
      DEBUG:__main__:Checking with includes ['fftw3.h']...no
      error: Could not find the FFTW header 'fftw3.h'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyfftw
Failed to build pyfftw
ERROR: Could not build wheels for pyfftw, which is required to install pyproject.toml-based projects
```
```
latha070% pip install git+https://github.com/haddocking/disvis.git@v2.1
Defaulting to user installation because normal site-packages is not writeable
Collecting git+https://github.com/haddocking/disvis.git@v2.1
  Cloning https://github.com/haddocking/disvis.git (to revision v2.1) to /private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-req-build-5vwcups4
  Running command git clone --filter=blob:none --quiet https://github.com/haddocking/disvis.git /private/var/folders/dk/ppqdqs693_n2ycpg6_y6vpg40000gp/T/pip-req-build-5vwcups4
  Resolved https://github.com/haddocking/disvis.git to commit a922bd079b41ad5ef3ac33f4e68968f8978626d2
  Preparing metadata (setup.py) ... done
Requirement already satisfied: numpy in /Users/latha070/Library/Python/3.9/lib/python/site-packages (from disvis==2.0.0) (1.23.4)
Building wheels for collected packages: disvis
  Building wheel for disvis (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [86 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-10.9-universal2-3.9
      creating build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/elements.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/pyclfft.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/__init__.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/rotations.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/pdb.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/volume.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/kernels.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/main.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/helpers.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      copying disvis/disvis.py -> build/lib.macosx-10.9-universal2-3.9/disvis
      creating build/lib.macosx-10.9-universal2-3.9/disvis/IO
      copying disvis/IO/mmcif.py -> build/lib.macosx-10.9-universal2-3.9/disvis/IO
      copying disvis/IO/__init__.py -> build/lib.macosx-10.9-universal2-3.9/disvis/IO
      copying disvis/IO/pdb.py -> build/lib.macosx-10.9-universal2-3.9/disvis/IO
      copying disvis/IO/mrc.py -> build/lib.macosx-10.9-universal2-3.9/disvis/IO
      creating build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u7111.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c600vc.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u27.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c600v.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u1201.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u5879.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u83.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/E.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48n309.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48n527.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u2219.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u1641.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u181.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u815.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u3733.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48n9.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u2947.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u1.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u8649.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u1153.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/data/c48u4749.npy -> build/lib.macosx-10.9-universal2-3.9/disvis/data
      copying disvis/kernels.cl -> build/lib.macosx-10.9-universal2-3.9/disvis
      running build_ext
      building 'disvis.libdisvis' extension
      creating build/temp.macosx-10.9-universal2-3.9
      creating build/temp.macosx-10.9-universal2-3.9/src
      clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration -I/Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -c src/libdisvis.c -o build/temp.macosx-10.9-universal2-3.9/src/libdisvis.o -ffast-math
      In file included from src/libdisvis.c:775:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/arrayobject.h:5:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/ndarraytypes.h:1948:
      /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
      #warning "Using deprecated NumPy API, disable it with " \
       ^
      1 warning generated.
      In file included from src/libdisvis.c:775:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/arrayobject.h:5:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/ndarraytypes.h:1948:
      /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
      #warning "Using deprecated NumPy API, disable it with " \
       ^
      1 warning generated.
      clang -bundle -undefined dynamic_lookup -arch arm64 -arch x86_64 -Wl,-headerpad,0x1000 build/temp.macosx-10.9-universal2-3.9/src/libdisvis.o -o build/lib.macosx-10.9-universal2-3.9/disvis/libdisvis.cpython-39-darwin.so
      building 'disvis._extensions' extension
      clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration -I/Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -c src/_extensions.c -o build/temp.macosx-10.9-universal2-3.9/src/_extensions.o -ffast-math
      In file included from src/_extensions.c:2:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/arrayobject.h:5:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
      In file included from /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/ndarraytypes.h:1948:
      /Users/latha070/Library/Python/3.9/lib/python/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
      #warning "Using deprecated NumPy API, disable it with " \
       ^
      src/_extensions.c:193:5: error: implicit declaration of function 'PyArray_XDECREF_ERR' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
          PyArray_XDECREF_ERR(py_out);
          ^
      src/_extensions.c:208:12: error: implicit declaration of function 'Py_InitModule' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
          (void) Py_InitModule("_extensions", mymethods);
                 ^
      src/_extensions.c:210:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
      };
      ^
      2 warnings and 2 errors generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for disvis
  Running setup.py clean for disvis
Failed to build disvis
ERROR: Could not build wheels for disvis, which is required to install pyproject.toml-based projects
```

I have previously installed numpy (v1.23.4) and Cython (v0.29.35) with pip. 

Thanks for your help!!

Cheers,
Mike

**Comment by rvhonorato** [comment]
Thanks again @michaellatham77, it seems that both of us overlooked the instructions 😅 

https://github.com/haddocking/disvis/blob/a922bd079b41ad5ef3ac33f4e68968f8978626d2/README.md?plain=1#L16-L18

When you install using python2.7 it all goes smoothly, see below my example using pyenv+virtualenv, you can probably achieve the same with anaconda.

```bash
# install python 2.7 with pyenv
$ pyenv install 2.7

# use it to install virtualenv
$ pyenv shell 2.7
$ pip install virtualenv

# create a virtual environment with python2.7
$ virtualenv -p `which python` .py2venv

# activate the virtual environment
$ source activate .py2venv/bin/activate

# install DisVis
(.py2venv) $ pip install numpy pyfftw
(.py2venv) $ pip install git+https://github.com/haddocking/disvis.git@v2.1

# execute it
(.py2venv) $ disvis -h
```
